### PR TITLE
BUG: Fix to_json when converting Period column

### LINKED ISSUE #31917: `to_json` of Series with period dtype results in AttributeError
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
test = pd.Series(pd.period_range('1/1/2011', freq='B', periods=3))
test.to_json(None, orient='table')
```

Leads to:

```
~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/core/generic.py in to_json(self, path_or_buf, orient, date_format, double_precision, force_ascii, date_unit, default_handler, lines, compression, index, indent)
   2362             compression=compression,
   2363             index=index,
-> 2364             indent=indent,
   2365         )
   2366

~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/io/json/_json.py in to_json(path_or_buf, obj, orient, date_format, double_precision, force_ascii, date_unit, default_handler, lines, compression, index, indent)
     83         default_handler=default_handler,
     84         index=index,
---> 85         indent=indent,
     86     ).write()
     87

~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/io/json/_json.py in __init__(self, obj, orient, date_format, double_precision, ensure_ascii, date_unit, index, default_handler, indent)
    289             raise ValueError(msg)
    290
--> 291         self.schema = build_table_schema(obj, index=self.index)
    292
    293         # NotImplemented on a column MultiIndex

~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/io/json/_table_schema.py in build_table_schema(data, index, primary_key, version)
    251     if data.ndim > 1:
    252         for column, s in data.items():
--> 253             fields.append(convert_pandas_type_to_json_field(s))
    254     else:
    255         fields.append(convert_pandas_type_to_json_field(data))

~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/io/json/_table_schema.py in convert_pandas_type_to_json_field(arr, dtype)
    115         field["ordered"] = ordered
    116     elif is_period_dtype(arr):
--> 117         field["freq"] = arr.freqstr
    118     elif is_datetime64tz_dtype(arr):
    119         if hasattr(arr, "dt"):

~/miniconda3/envs/qgrid/lib/python3.7/site-packages/pandas/core/generic.py in __getattr__(self, name)
   5272             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5273                 return self[name]
-> 5274             return object.__getattribute__(self, name)
   5275
   5276     def __setattr__(self, name: str, value) -> None:

AttributeError: 'Series' object has no attribute 'freqstr'
```

This issue also crops up in dataframes where a column is a period Series and `orient='table'`. 

#### Problem description

I can't find anything that says this would be unsupported. 

#### Expected Output

Json string of series representation. 

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.1
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.2.0.post20200210
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.12.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by colonesej**
AttributeError  when trying to access frequency string ``freqstr`` directly on Series of ``Period`` type. 

Modified to get it through ``dtype`` of Series, which is known to be ``Period``.

- [X] closes #31917
- [X] fix #31917
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

```python
import pandas as pd
test = pd.Series(pd.period_range('1/1/2011', freq='B', periods=3))
test.to_json(None, orient='table')

>>> '{"schema":{"fields":[{"name":"index","type":"integer"},{"name":"values","type":"datetime","freq":"B"}],"primaryKey":["index"],"pandas_version":"0.20.0"},"data":[{"index":0,"values":{"day":3,"dayofyear":3,"daysinmonth":31,"freqstr":"B","is_leap_year":false,"month":1,"ordinal":10697,"qyear":2011,"start_time":"2011-01-03T00:00:00.000Z","week":1,"weekofyear":1}},{"index":1,"values":{"day":4,"dayofyear":4,"daysinmonth":31,"freqstr":"B","is_leap_year":false,"month":1,"ordinal":10698,"qyear":2011,"start_time":"2011-01-04T00:00:00.000Z","week":1,"weekofyear":1}},{"index":2,"values":{"day":5,"dayofyear":5,"daysinmonth":31,"freqstr":"B","is_leap_year":false,"month":1,"ordinal":10699,"qyear":2011,"start_time":"2011-01-05T00:00:00.000Z","week":1,"weekofyear":1}}]}'
```

### CODE REVIEW

**Comment by WillAyd** [comment]
Can you add a test for this? @jbrockmendel

**Review by jreback — CHANGES_REQUESTED** [review]
needs a test replicating the OP & a whatsnew note (bug fixes in 1.1 / IO)

**Comment by WillAyd** [comment]
Looks like a merge conflict in the whatsnew - can you fix and add a test case in pandas/tests/io/json/test_pandas.py?

**Comment by colonesej** [comment]
OK, I'll do it, but first I need to figure it out how to make a test. Have to read more carefully the guide.

**Comment by WillAyd** [comment]
@colonesej you should just be able to place a test in pandas/tests/io/json/test_pandas.py for this (and can look at other examples there too)

Can you fix up merge conflict here as well?

**Comment by colonesej** [comment]
I made this roundtrip test completely based on the one for a time series with a DatetimeIndex. Some points: 

*  I could not find any option to make `read_json` treat the data and generate a PeriodIndex when `orient` equals to `"column" or "index"`. Conversion does not carry frequency information. I'm not sure what exactly should be the behaviour so just converted manually to Period and inferred freq is OK.
* When `orient = "split"`, the test fails as `read_json` cannot convert back the dict created by `to_json`. It works with a DatetimeIndex, which conversion results in float formatted dates. With Period it converts to separated keys for different aspects of Period object. I could not find where these were determined nor what should be preferred.

To illustrate, using `series.to_json(orient='split')` when index is:

DatetimeIndex:

>>> '{"name":null,"index":[946857600000,946944000000,947030400000,947116800000,947203200000,947462400000,947548800000,947635200000,947721600000,947808000000,948067200000,948153600000,948240000000,948326400000,948412800000,948672000000,948758400000,948844800000,948931200000,949017600000,949276800000,949363200000,949449600000,949536000000,949622400000,949881600000,949968000000,950054400000,950140800000,950227200000],"data":[-0.7273489893,0.43763153,0.0064352361,-1.0478282669,0.6605651751,0.2847295244,1.6006677017,-1.2731074841,0.3004426454,0.274394972,0.0402248331,1.5284606656,0.5741635466,-0.3126890855,1.4765392606,0.3663030854,0.3132193099,-0.0730319057,-0.3956868944,1.4626210011,0.2129331654,-0.6671879793,-1.6477379639,0.7626651704,-1.7852083837,0.0707591559,1.3089143648,-1.0282706187,-0.7853299219,-0.2119391625]}'

PeriodIndex:

>>>'{"name":null,"index":[{"day":3,"dayofyear":3,"daysinmonth":31,"freqstr":"B","is_leap_year":true,"month":1,"ordinal":7827,"qyear":2000,"start_time":946857600000,"week":1,"weekofyear":1},

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -529,6 +529,7 @@ I/O
 - Bug in :meth:`DataFrame.to_sql` where an ``AttributeError`` was raised when saving an out of bounds date (:issue:`26761`)
 - Bug in :meth:`read_excel` did not correctly handle multiple embedded spaces in OpenDocument text cells. (:issue:`32207`)
 - Bug in :meth:`read_json` was raising ``TypeError`` when reading a list of booleans into a Series. (:issue:`31464`)
+- Bug in :meth:`to_json` was raising `AttributeError` with column or Series of type `Period` (:issue:`31917`).
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.1.0.rst:532]: "type `Period`" -> "``PeriodDtype``", no period at the end, double-ticks around AttributeError

File: pandas/conftest.py [hunk 2]
```diff
@@ -473,6 +473,15 @@ def _create_series(index):
 }
 
 
+@pytest.fixture
+def period_series():
```
Comment (jbrockmendel) [inline pandas/conftest.py:477]: @WillAyd did we settle on a naming scheme for these?

Comment (WillAyd) [reply pandas/conftest.py:477]: Seems reasonable; we already have a datetime_series so this follows that pattern

File: pandas/tests/io/json/test_pandas.py [hunk 3]
```diff
@@ -650,6 +650,21 @@ def test_series_roundtrip_timeseries(self, orient, numpy, datetime_series):
 
         tm.assert_series_equal(result, expected)
 
+    def test_series_roundtrip_periodseries(self, orient, period_series):
+        # GH32665: Fix to_json when converting Period column/series
+        data = period_series.to_json(orient=orient)
+        result = pd.read_json(data, typ="series", orient=orient)
```
Comment (WillAyd) [inline pandas/tests/io/json/test_pandas.py:656]: This issue affects reading to a DataFrame as well right? If so can you test that?

Comment (colonesej) [reply pandas/tests/io/json/test_pandas.py:656]: OK, I can do that.

**Comment by colonesej** [comment]
The point with `orient="split"` seems to be connected to how `to_json` converts columns/series with values of `PeriodDtype`. It constructs that dictionary that cannot be converted back by `read_json`. Not directly.

I don't know what should be the behaviour here: 
* change read_json to specifically treat when `orient="split"` expecting a dict of attributes, or
* change to_json to convert Period values to float (datetime behaviour) or use `__repr__`. The former 
is what other options of `to_json` seems to use.

I made this other test with a frame column with `PeriodDtype` values and eventually test this by the time a solution is proposed.

**Comment by WillAyd** [comment]
@colonesej can you merge master and try to get green?

**Comment by colonesej** [comment]
Yes, but you mean, make the roundtrip tests I've made pass? 

I could have some help about how `to_json` works with `orient = split`.

Or could make a different test, testing based on some expected output template. Will not solve the roundtrip issue but will make it green. What is preferable?

**Comment by WillAyd** [comment]
If there’s one orient that doesn’t work you can pytest.skip it for now - check some of the other tests in the module for examples

**Comment by WillAyd** [comment]
@colonesej is this PR still active?

**Pull request closed without merging**